core/sile.lua

```diff
@@ -240,10 +241,251 @@ function SILE.resolveFile(filename, pathprefix)
 
+SILE.currentCommandStack = {
+  --- Internal: push-pop balance checking ID
+  _lastPushId = 0,
+  --- Internal: Function assigned to frames to convert them to string
+  _frameToString = function(self, skipFile, withAttrs)
+    local str
+    if skipFile or not self.file then
+      str = ""
+    else
+      str = self.file .. " "
+    end
+
+    if self.line then
+      str = str .. "l." .. self.line .. " "
+      if self.column then
+        str = str .. "col." .. self.column .. " "
+      end
+    end
+
+    if self.tag then
+      -- Command
+      str = str .. "\\" .. self.tag
+      if withAttrs then
+        str = str .. "["
+        local first = true
+        for key, value in pairs(self.options) do
+          if first then first = false else str = str .. ", " end
+          str = str .. key .. "=" .. value
+        end
+        str = str .. "]"
+      end
+    elseif self.text then
+      -- Literal string
+      local text = self.text
+      if text:len() > 20 then
+        text = text:sub(1, 18) .. "…"
+      end
+      text = text:gsub("\n", "␤"):gsub("\t", "␉"):gsub("\v", "␋")
+
+      str = str .. "\"" .. text .. "\""
+    else
+      -- Unknown
+      str = str .. type(self.content) .. "(" .. self.content .. ")"
+    end
+
+    --str = str .. "\n\n" .. self.pushTrace .. "\n"
+
+    return str
+  end
+}
+--- Push a command frame on to the stack to record the execution trace for debugging.
+--- Carries information about the command call, not the command itself.
+--- Must be popped with `pop(returnOfPush)`.
+function SILE.currentCommandStack:pushCommand(tag, line, column, options, file)
+  return self:_pushFrame({
+    tag = tag or "???",
+    file = file or SILE.currentlyProcessingFile,
+    line = line,
+    column = column,
+    options = options or {}
+  })
+end
+--- Push a command frame on to the stack to record the execution trace for debugging.
+--- Command arguments are inferred from AST content, any item may be overridden.
+--- Must be popped with `pop(returnOfPush)`.
+function SILE.currentCommandStack:pushContent(content, tag, line, column, options, file)
+  return self:_pushFrame({
+    tag = tag or (type(content) == "table" and content.tag) or "???",
+    file = file or (type(content) == "table" and content.file) or SILE.currentlyProcessingFile,
+    line = line or (type(content) == "table" and content.line),
+    column = column or (type(content) == "table" and content.col),
+    options = options or (type(content) == "table" and content.attr) or {}
+  })
+end
+--- Push a text that is going to get typeset on to the stack to record the execution trace for debugging.
+--- Must be popped with `pop(returnOfPush)`.
+function SILE.currentCommandStack:pushText(text, line, column, file)
+  return self:_pushFrame({
+    text = text,
+    file = file,
+    line = line,
+    column = column
+  })
+end
+--- Internal: Push complete frame
+function SILE.currentCommandStack:_pushFrame(frame)
+  frame.toString = self._frameToString
```

Comment:
Lua already has a `tostring()` function as well as `__tostring()` methods on tables. Naming this `toString()` is confusing — if you need to pass arguments then at least call it `.dump()` or something like that, but I would try to standardize on a `__tostring()` that takes no arguments and if you need a special output beyond that build it on top of that.

Reply:
I don't want to call it `dump`, because that would imply (at least to me) that it is intended as a debug info. But this actually formats an user-facing string. I can't think of a better way to call this, as it is a method on frame, which converts it into a string. But it could be confused with `tostring`, so would `toFrameString` be better?

Reply:
Why do you call it "user facing" and contrast that with "debug info"? The only place this function is even called currently is in the context of `SU.debugging("commandStack")`, so that seems like a false dichotomy at this point. Did you have another use case in mind?

Reply:
No, the `frame.toString` is the main source of the error trace information. The internal architecture is:
- There are stack frames on `currentCommandStack`. Those contain information about some call/command/text which is being processed.
- Each stack frame also has a `toString` method, for convenience.
- When a short or long location message is generated through `locationInfo` or `locationTrace`, `_locationTraceEntry` is called for relevant stack traces. This looks at single stack frame (+ some other frames to generate "after" and "near" contextual info), converts it to string via `toString` and pieces it together with other contextual info.
- For `locationInfo`, the output of `_locationTraceEntry` is enough.
- For `locationTrace`, `toString` is also called for other frames on the stack, to capture all information.

`toString` is what converts collected information (for example from `content` AST), into something like: "fontbug.sil l.6 col.1 \verbatim".

Reply:
Thanks for the explanation. This makes a _LOT_ more sense if your most recent code, and I'm following that a lot better now.

The one concern I have left (and based on this explanation I now realize was a large contributor to by confusion) is that `frame` might be perfectly appropriate for stack tracer **except** in the content of a typesetting application that handles everything with in frames. I think we should universally stick to "frame" meaning a content frame on a page across the code base and use something else for whatever unit of content you are tracing here.

Reply:
...I can't come up with anything better so I'm actually going to tie off this issue too unless somebody has a brilliant suggestion.